analytics/location_heatmaps/geo_utils.py

```diff
@@ -278,141 +296,124 @@ def split_regions(tree_prefix_list,
-          unchanged += 1
-          new_tree[f'{neg_prefix}'] = len(new_tree_prefix_list)
-          new_tree_prefix_list.append(f'{neg_prefix}')
-          new_tree[f'{pos_prefix}'] = len(new_tree_prefix_list)
-          new_tree_prefix_list.append(f'{pos_prefix}')
-  else:
-    for i in range(len(tree_prefix_list)):
-      if expand_all:
-        count = threshold + 1
-      else:
-        if count_min:
-          count = count_min.query(tree_prefix_list[i])
-        else:
-          count = vector_counts[i]
-      prefix = tree_prefix_list[i]
-
-      # check whether the tree has reached the bottom
-      if len(prefix.split('/')) >= image_bit_level:
-        continue
-      if last_result is not None:
-        (last_prefix, last_prefix_pos) = last_result.tree.longest_prefix(prefix)
-        if last_prefix is None:
-          cond = False
-        else:
-          last_count = last_result.sum_vector[last_prefix_pos]
-          p = (last_count - count) / last_count
-          if p <= 0 or count < 5 or last_count < 5:
-            cond = False
-            # print(last_prefix, prefix, last_prefix_pos, last_count,
-            #       count)
-          else:
-            conf_int = 1.96 * np.sqrt((p * (1 - p) / last_count)) * last_count
-            cond = conf_int < threshold
-            intervals.append(conf_int)
-            # print(last_prefix, prefix, last_prefix_pos, last_count, count, conf_int, cond)
+  for i in range(len(tree_prefix_list)):
+    if count_min:
+      count = count_min.query(tree_prefix_list[i])
+    else:
+      count = vector_counts[i] if not expand_all else np.inf
+    prefix = tree_prefix_list[i]
+
+    # check whether the tree has reached the bottom
+    if len(prefix.split('/')) >= image_bit_level:
+      continue
+    if last_result:
+      cond = create_confidence_interval_condition(last_result, prefix, count,
+                                                  split_threshold)
+    else:
+      cond = count > split_threshold
+    if expand_all or cond:
+      for child in DEFAULT_CHILDREN:
+        new_prefix = f'{prefix}/{child}'
+        fresh_expand += update_tree(new_prefix, new_tree, new_tree_prefix_list)
+    else:
+      if collapse_threshold is not None and \
+        count <= collapse_threshold and len(prefix) > 2:
+        old_prefix = prefix[:-3]
+        collapsed += 1
+        created += update_tree(old_prefix, new_tree, new_tree_prefix_list)
       else:
-        cond = count > threshold
-      # print(cond, threshold, count)
-      if cond:
-        for child in DEFAULT_CHILDREN:
-          new_prefix = f'{prefix}/{child}'
-          if not new_tree.has_key(new_prefix):
-            fresh_expand += 1
-            new_tree[new_prefix] = len(new_tree_prefix_list)
-            new_tree_prefix_list.append(new_prefix)
+        unchanged += update_tree(prefix, new_tree, new_tree_prefix_list)
+
+  return new_tree, new_tree_prefix_list, fresh_expand
+
+
+def split_regions_aux(tree_prefix_list,
```

Comment:
I suspect that even more of these two functions could be shared (in particular, the basic structure of looping over prefixes and adding nodes to the tree as appropriate for the splitting & collapsing criteria), but acknowledge that it may not actually improve readability much more to do further surgery. Please consider sharing that prefix-looping structure, but if you can't see a clean and easy way to do so, that's fine.

Reply:
yeah, I agree, it's just that I need to look at both bits in data that is hard to unify. Maybe once we go to multiple dimensions we can just unify everything.